daemon/appinfo.c

```diff
@@ -0,0 +1,747 @@
+                              g_variant_new_string(appinfo_get_name(self)));
+
+        g_variant_builder_add(builder, "{sv}", DESKTOP_KEY_TYPE,
+                              g_variant_new_string(appinfo_get_type(self)));
+
+        g_variant_builder_add(builder, "{sv}", DESKTOP_KEY_ICON,
+                              g_variant_new_string(appinfo_get_icon(self)));
+
+        g_variant_builder_add(builder, "{sv}", DESKTOP_KEY_EXEC,
+                              g_variant_new_string(appinfo_get_exec(self)));
+
+        g_variant_builder_add(builder, "{sv}", DESKTOP_KEY_NO_DISPLAY,
+                              g_variant_new_boolean(appinfo_get_no_display(self)));
+
+        /* Maemo properties
+         */
+        g_variant_builder_add(builder, "{sv}", MAEMO_KEY_SERVICE,
+                              g_variant_new_string(appinfo_get_service(self)));
+
+        g_variant_builder_add(builder, "{sv}", MAEMO_KEY_OBJECT,
+                              g_variant_new_string(appinfo_get_object(self)));
+
+        g_variant_builder_add(builder, "{sv}", MAEMO_KEY_METHOD,
+                              g_variant_new_string(appinfo_get_method(self)));
+
+        /* Sailjail properties
+         */
+        g_variant_builder_add(builder, "{sv}", SAILJAIL_KEY_ORGANIZATION_NAME,
+                              g_variant_new_string(appinfo_get_organization_name(self)));
+
+        g_variant_builder_add(builder, "{sv}", SAILJAIL_KEY_APPLICATION_NAME,
+                              g_variant_new_string(appinfo_get_application_name(self)));
+
+        g_variant_builder_add(builder, "{sv}", SAILJAIL_KEY_PERMISSIONS,
+                              stringset_to_variant(appinfo_get_permissions(self)));
+    }
+
+    GVariant *variant = g_variant_builder_end(builder);
+    g_variant_builder_unref(builder);
+
+    return variant;
+}
+
+gchar *
+appinfo_to_string(const appinfo_t *self)
+{
+    gchar    *string  = NULL;
+    GVariant *variant = appinfo_to_variant(self);
+    if( variant ) {
+        string = g_variant_print(variant, false);
+        g_variant_unref(variant);
+    }
+    return string;
+}
+
+/* ------------------------------------------------------------------------- *
+ * APPINFO_ATTRIBUTE
+ * ------------------------------------------------------------------------- */
+
+bool
+appinfo_valid(const appinfo_t *self)
+{
+    return appinfo_get_state(self) == APPINFO_STATE_VALID;
+}
+
+control_t *
+appinfo_control(const appinfo_t *self)
+{
+    return applications_control(appinfo_applications(self));
+}
+
+applications_t *
+appinfo_applications(const appinfo_t *self)
+{
+    return self->anf_applications;
+}
+
+const gchar *
+appinfo_id(const appinfo_t *self)
+{
+    return self->anf_appname;
+}
+
+/* ------------------------------------------------------------------------- *
+ * APPINFO_PROPERTY
+ * ------------------------------------------------------------------------- */
+
+static const char appinfo_unknown[] = "unknown";
```

Comment:
May be we should use some value that can't be used as a valid application name. I mean if it is possible to create unknown.rpm, then it will be indistinguishable with the case when no information is supplied.

Reply:
Hmm.. this should have been accompanied with FIXME note...

Basically I'm inclined to think that we should just omit values that are not defined. But that would mean that clients need to be prepared for such thing also when dealing with changes -> got unsure what to do about it -> used placeholder value.

[But that "unknown" string is there just because I was using qdbus for testing and it mangles reply output badly enough to be unintelligible if message contains empty strings.]

Reply:
Changed it so that: if desktop file does not define a string value, placeholder string is used internally within daemon but value is omitted from data set sent over dbus -> client can deal with the situation as appropriate for the situation.